stl/inc/bitset

```diff
@@ -347,13 +347,13 @@ public:
     _NODISCARD _CONSTEXPR23 basic_string<_Elem, _Tr, _Alloc> to_string(
         const _Elem _Elem0 = static_cast<_Elem>('0'), const _Elem _Elem1 = static_cast<_Elem>('1')) const {
         // convert bitset to string
-        basic_string<_Elem, _Tr, _Alloc> _Str(_Bits, _Elem0);
-        for (size_t _Pos = 0; _Pos < _Bits; ++_Pos) {
-            if (_Subscript(_Bits - 1 - _Pos)) {
-                _Str[_Pos] = _Elem1;
+        basic_string<_Elem, _Tr, _Alloc> _Str;
+        _Str._Resize_and_overwrite(_Bits, [this, _Elem0, _Elem1](_Elem* _Buf, size_t _Len) {

```

Comment:
No change requested: I observed that we could ignore `_Len` and take advantage of our knowledge that it will always be the compile-time constant `_Bits`, which is available without capturing. I checked the codegen, and this indeed allows the compiler to bake in the constant. However, while benchmarking showed some improvement for `BM_bitset_to_string<64, char>`, it showed significant pessimization for `BM_bitset_to_string<15, char>` .

```
BEFORE USING _Bits
------------------------------------------------------------------------------------
Benchmark                                          Time             CPU   Iterations
------------------------------------------------------------------------------------
BM_bitset_to_string<15, char>                    140 ns          144 ns      4977778
BM_bitset_to_string<64, char>                   3336 ns         3348 ns       224000
BM_bitset_to_string_large_single<char>          2008 ns         2040 ns       344615
BM_bitset_to_string<7, wchar_t>                 88.3 ns         87.9 ns      7466667
BM_bitset_to_string<64, wchar_t>                2211 ns         2197 ns       298667
BM_bitset_to_string_large_single<wchar_t>       1451 ns         1430 ns       448000

AFTER USING _Bits
------------------------------------------------------------------------------------
Benchmark                                          Time             CPU   Iterations
------------------------------------------------------------------------------------
BM_bitset_to_string<15, char>                    640 ns          642 ns      1120000
BM_bitset_to_string<64, char>                   2646 ns         2609 ns       263529
BM_bitset_to_string_large_single<char>          2010 ns         2040 ns       344615
BM_bitset_to_string<7, wchar_t>                 82.3 ns         82.0 ns      8960000
BM_bitset_to_string<64, wchar_t>                2407 ns         2441 ns       320000
BM_bitset_to_string_large_single<wchar_t>       1372 ns         1360 ns       448000
```

Reply:
I think despite some pessimization of the certain case, it is still the right thing to do

Reply:
Oh, actually the pessimization is too much. I think, angry loop unrolling strikes here too.

Reply:
Let's keep it as a parameter for now then

Reply:
As `_Bits` is the natural decision but we have to use `_Len` for performance reasons, I think we need to add a comment for it.

Reply:
Hmm, how to compare benchmark results before and after change in the same run in a convenient way?

<details>
<summary>Benchmark</summary>

```cpp
#include <array>
#include <benchmark/benchmark.h>
#include <bit>
#include <bitset>

#include <climits>
#include <cstddef>
#include <cstdint>
#include <random>
#include <iostream>

using namespace std;

struct impl_old {
    template <class _Elem = char, size_t _Bits, class _Tr = char_traits<_Elem>, class _Alloc = allocator<_Elem>>
    static basic_string<_Elem, _Tr, _Alloc> to_string(const std::bitset<_Bits>& bs,
        const _Elem _Elem0 = static_cast<_Elem>('0'), const _Elem _Elem1 = static_cast<_Elem>('1')) {
        basic_string<_Elem, _Tr, _Alloc> _Str(_Bits, _Elem0);
        for (size_t _Pos = 0; _Pos < _Bits; ++_Pos) {
            if (bs[_Bits - 1 - _Pos]) {
                _Str[_Pos] = _Elem1;
            }
        }
        return _Str;
    }
};

struct impl_new_Bits {
    template <class _Elem = char, size_t _Bits, class _Tr = char_traits<_Elem>, class _Alloc = allocator<_Elem>>
    static basic_string<_Elem, _Tr, _Alloc> to_string(const std::bitset<_Bits>& bs,
        const _Elem _Elem0 = static_cast<_Elem>('0'), const _Elem _Elem1 = static_cast<_Elem>('1')) {
        basic_string<_Elem, _Tr, _Alloc> _Str;
        _Str.resize_and_overwrite(_Bits, [&bs, _Elem0, _Elem1](_Elem* _Buf, size_t) {
            for (size_t _Pos = 0; _Pos < _Bits; ++_Pos) {
                _Buf[_Pos] = bs[_Bits - 1 - _Pos] ? _Elem1 : _Elem0;
            }
            return _Bits;
            });
        return _Str;
    }
};

struct impl_new_Len {
    template <class _Elem = char, size_t _Bits, class _Tr = char_traits<_Elem>, class _Alloc = allocator<_Elem>>
    static basic_string<_Elem, _Tr, _Alloc> to_string(const std::bitset<_Bits>& bs,
        const _Elem _Elem0 = static_cast<_Elem>('0'), const _Elem _Elem1 = static_cast<_Elem>('1')) {
        basic_string<_Elem, _Tr, _Alloc> _Str;
        _Str.resize_and_overwrite(_Bits, [&bs, _Elem0, _Elem1](_Elem* _Buf, size_t _Len) {
            for (size_t _Pos = 0; _Pos < _Len; ++_Pos) {
                _Buf[_Pos] = bs[_Len - 1 - _Pos] ? _Elem1 : _Elem0;
            }
            return _Len;
            });
        return _Str;
    }
};

namespace {
    const auto random_bits = [] {
        mt19937_64 rnd{};
        array<uint64_t, 32> arr;
        for (auto& d : arr) {
            d = rnd();
        }
        return arr;
        }();

        template <class op, size_t N, class charT>
        void BM_bitset_to_string(benchmark::State& state) {
            for (auto _ : state) {
                for (const auto& bits : random_bits) {
                    bitset<N> bs{ bits };
                    benchmark::DoNotOptimize(op::template to_string<charT>(bs));
                }
            }
        }

        template <class op, class charT>
        void BM_bitset_to_string_large_single(benchmark::State& state) {
            const auto large_bitset = bit_cast<bitset<CHAR_BIT * sizeof(random_bits)>>(random_bits);
            for (auto _ : state) {
                benchmark::DoNotOptimize(op::template to_string<charT>(large_bitset));
            }
        }
} // namespace

#define MAKE_BENCHMARKS(impl)\
BENCHMARK(BM_bitset_to_string<impl, 15, char>)->Name("\nBM_bitset_to_string<" #impl "_old, 15, char>");\
BENCHMARK(BM_bitset_to_string<impl, 64, char>);\
BENCHMARK(BM_bitset_to_string_large_single<impl, char>);\
BENCHMARK(BM_bitset_to_string<impl, 7, wchar_t>);\
BENCHMARK(BM_bitset_to_string<impl, 64, wchar_t>);\
BENCHMARK(BM_bitset_to_string_large_single<impl, wchar_t>);

MAKE_BENCHMARKS(impl_old);
MAKE_BENCHMARKS(impl_new_Bits);
MAKE_BENCHMARKS(impl_new_Len);

BENCHMARK_MAIN();
```
</details>

(And I'm feeling nervious that it seems the effect is very complex and can even vary between different runs)
![image](https://github.com/microsoft/STL/assets/60953653/0d63c980-06ee-46f8-8f77-bef57d5af950)
```
---------------------------------------------------------------------------------------------------
Benchmark                                                         Time             CPU   Iterations
---------------------------------------------------------------------------------------------------

BM_bitset_to_string<impl_old_old, 15, char>                   1531 ns         1549 ns       373333
BM_bitset_to_string<impl_old, 64, char>                        8230 ns         8231 ns       112000
BM_bitset_to_string_large_single<impl_old, char>               6477 ns         6487 ns       149333
BM_bitset_to_string<impl_old, 7, wchar_t>                       642 ns          638 ns       857710
BM_bitset_to_string<impl_old, 64, wchar_t>                     7436 ns         7533 ns       112000
BM_bitset_to_string_large_single<impl_old, wchar_t>            4339 ns         4325 ns       144516

BM_bitset_to_string<impl_new_Bits_old, 15, char>              1156 ns         1172 ns       560000
BM_bitset_to_string<impl_new_Bits, 64, char>                   6346 ns         6452 ns        89600
BM_bitset_to_string_large_single<impl_new_Bits, char>          2344 ns         2354 ns       278756
BM_bitset_to_string<impl_new_Bits, 7, wchar_t>                  210 ns          207 ns      4824615
BM_bitset_to_string<impl_new_Bits, 64, wchar_t>                6676 ns         6696 ns        74667
BM_bitset_to_string_large_single<impl_new_Bits, wchar_t>       4516 ns         4450 ns       154483

BM_bitset_to_string<impl_new_Len_old, 15, char>                251 ns          254 ns      3200000
BM_bitset_to_string<impl_new_Len, 64, char>                    5659 ns         5755 ns        89600
BM_bitset_to_string_large_single<impl_new_Len, char>           3870 ns         3850 ns       280000
BM_bitset_to_string<impl_new_Len, 7, wchar_t>                   255 ns          255 ns      3733333
BM_bitset_to_string<impl_new_Len, 64, wchar_t>                15309 ns        15415 ns        72983
BM_bitset_to_string_large_single<impl_new_Len, wchar_t>        7019 ns         7150 ns        89600
```

Reply:
Are there problems with my comparing benchmark? I'm finding `resize_and_overwrite`&`_Len` approach stably slow for large wchar_t cases in my testing.

Reply:
It may be CPU-dependent. 
For `_Bits` the compiler is more likely to advantage of knowing `_Bits` and unroll loops or vectorize something.
For `_Len` the compiler may be restricted from doing that.

Whether the optimizations like loop unrolling are useful or harmful -- dependent on CPU type.

Also non-unrolled loop (`_Len`) is subject to JCC errata pessimization on some CPUs, `/QIntel-jcc-erratum` might help in this case https://learn.microsoft.com/ru-ru/cpp/build/reference/qintel-jcc-erratum?view=msvc-170

I'm neutral with regards to `_Bits` vs `_Len` if none of them are absolutely advantage.

Reply:
I tend to believe that by using `_Bits` we can get general improvements over previous impl, though not significant in some cases like `small_bitset.to_string<char>`. By using `_Len` we are introducing some trade-offs.

Reply:
I suggest don't think too much on `_Bits` vs `_Len`. The next step would be vectorization, it would be obviously faster than this, and obviously will use `_Bits`.